### ISSUE TITLE
Issues when combined with (global) mixins

### ISSUE THREAD

**Issue by jvv** [GuillaumeLeclerc/vue-google-maps#37]
I'm trying to include this component into my project, but every time I load the / a map I get an error:

Uncaught TypeError: Cannot read property 'accessToken' of undefined

Where accessToken is defined in data object from my global mixins file (where I basically have a simple vue resource http call function I use throughout my project).

I checked to see if I could get a map working in a different environment (eg: without mixins) and this seems to not be an issue at all.

Is the advice to re-factor my mixins - to not use that at all, or is something else going on?

code snippets: http://pastebin.com/s0aLxXAV

**Comment by smartpierre** [comment]
As specified in the documentation It is not adviced to use global mixins as it will apply to every component in your project.
I apply mixins component by component and I doesn't make any issues.

http://vuejs.org/guide/mixins.html#Global-Mixin

**Comment by jvv** [comment]
As I didn't want to repeat myself constantly, the fact that it will apply to all my components was the part which I really enjoyed :-) Unfortunately this won't work with the vue maps, so I will have to start including a mixin file in all my components and try again. Thanks for the advice :+1:

**Comment by smartpierre** [comment]
Yeah I do understand ! But it is hazardous to apply code to EVERY component. You better add one line on each of your components and be sure where your mixins will be merged, it's not that much really :) .

**Comment by GuillaumeLeclerc** [comment]
I took a look at your problem and I don't see how it could be a problem related with `vue-google-maps` It just looks like a JavaScript error (accessing an undefined variable). Can you show us more about your code ? I would especially interested in the place where you call `load`

Moreover I have a question: Why do you want to load `vue-google-maps` in a mixin. That means you will load it (and do you query to get the accessToken) multiple times. This is not required. You only need to do it once. I really don't get why you are doing it here but I must have missed something.

**Comment by jvv** [comment]
Thanks for looking at the issue - I think I did not explain properly though - I don't load `vue-google-maps` in the mixin, but in a component affected by the mixin. 

My setup is as follow:

- Main.vue 
-- router-view
--- Explore.vue < loads `vue-google-maps`
--- And other .vue files

Explore.vue => http://pastebin.com/QEyBtdtj (very basic, only want to show a map currently)

In the vue files where I don't load `vue-google-maps` I don't see any errors regarding the global accessToken, only in Explore.vue

`Uncaught TypeError: Cannot read property 'accessToken' of undefined`

---------------------

Now, I did try some other stuff. In the Explore.vue from pastebin I have a component object set up, loading the Map. If I remove this I don't see the error anymore (and my explore component has an accessToken) but also no map.  

The HTML I see in devtools for the map however does something strange;

`<map center="[object Object]" zoom="7"></map>`

With the component loaded - no map object is seen in sources (no title as well), but this is due to VueJS stopping on the empty accessToken issue.

Is there any more information I can provide? Or can you already see what it is I'm doing wrong.

**Comment by GuillaumeLeclerc** [comment]
It seems your problem is pretty simple and neither related to mixins nor `vue-gogle-maps`.

If I understood well you have a method able to get an access token asynchronously (with an AJAX request), this is in a mixin. Then you load this mixin in an `Explore.vue` that loads `vue-google-maps` with the access you got. 

Am I right ? If yes I'll explain you what's wrong if not can you rephrase your problem please ?

**Comment by jvv** [comment]
You described the setup correct :-)

**Comment by GuillaumeLeclerc** [comment]
So this is what's happening:

- Your `Explorer.vue` component is loaded and try to load `vue-google-maps` with a token.
- The token should be in your mixin
- Your mixin triggers an Ajax query that will return the access token LATER
- The miethods needs to return something (but the query hasn't event started because we are in the same call stack
- Your method has to return `undefined` because your Ajax call will be done in something like 300ms but JavaScript calls are __non-blocking__ so your code must continue
- Since your methods returns undefined you have this error

Am I guessing right ? I can't really check this with the little amount of code you gave us.

**Comment by jvv** [comment]
Hm, not quite I think.

- Explore.vue component loads, triggers the mixin in which I have a ready function
- Global mixin checks if there already is a token (cookie), if not it will trigger an ajax call which generates the token
- There currently is no call/method in Explore.vue requiring the token (but I will add later, get markers and such - these should of course only run when there is a token available)

99/100 times Explore will be loaded after a token is already requested from the server, so no ajax call would be required in those cases. 

Also, when I add a (valid) accessToken to the data object within the mixin (so no check or call at all would be needed) I have the same result. 

----------

From the docs, section `Or use the power of Vue.js within a google map like this:` there is no component object with (eg) Maps, Marker and such. I only get the accessToken error when I do have the component object in my code, without I have the `<map center="[object Object]" zoom="7"></map>` - where the object is from my data `{ center  : { lat: 10, lng: 10 } }`

Really appreciate the time you spend trying to help me. If it would help, I could put some more code online - but have to cut out some sensitive data (api stuff) first.

**Comment by GuillaumeLeclerc** [comment]
I think I'm lost now because in your last message you say : "there is no call in Explore.vue requiring the token" but in your fiddle you have `load('-my-gmap-token')`.

I don't think I will be able to help with more code sorry.

**Comment by jvv** [comment]
I thought you meant the accessToken I require to do my AJAX calls. I have the gmap token inside the `load('')` hardcoded. No trigger to get the token from anywhere else.

**Comment by GuillaumeLeclerc** [comment]
So your problem is not related to with the google token ? Then I don't see how it is a `vue-google-maps` problem .

**Comment by GuillaumeLeclerc** [comment]
I think I'm starting to wrap my head around the problem. Did you tried the latest version I published yesterday ? (v0.1.13)

**Comment by jvv** [comment]
I added the 0.1.13 just now, did not seem to change anything for me. I did however changed something which made everything work. I thought it was after I added markers (and loading markers from server) but when I removed them the map still worked. 

Sorry to take your time @GuillaumeLeclerc - do you have a donate button somewhere? I'll buy you some beer.

**Comment by GuillaumeLeclerc** [comment]
I'm so glad it worked for you ! 

Of course if you like the library here is my link : https://paypal.me/GuillaumeLeclerc

**Comment by jvv** [comment]
Hope you'll enjoy the beer (or, whatever you'll do with it ;-))
